python/test/unit/fem/test_fem_pipeline.py

```diff
@@ -0,0 +1,134 @@
+        laplace_u = 0
+        for i in range(self.gdim):
+            laplace_u += sp.diff(self.u, self.x[i], self.x[i])
+        return laplace_u
+
+
+def boundary(x):
+    """Boundary marker """
+    condition = np.logical_or(x[:, 0] < 1.0e-6, x[:, 0] > 1.0 - 1.0e-6)
+    for dim in range(1, x.shape[1]):
+        c_dim = np.logical_or(x[:, dim] < 1.0e-6, x[:, dim] > 1.0 - 1.0e-6)
+        condition = np.logical_or(condition, c_dim)
+    return condition
+
+
+@pytest.mark.parametrize("degree", [2, 3, 4])
+@pytest.mark.parametrize("cell_type, gdim", [(CellType.interval, 1), (CellType.triangle, 2),
+                                             (CellType.quadrilateral, 2), (CellType.tetrahedron, 3),
+                                             (CellType.hexahedron, 3)])
+def test_manufactured_poisson(degree, cell_type, gdim):
+    """ Manufactured Poisson problem, solving u = Pi_{i=0}^gdim (1 - x[i]) * x[i]^(p - 1)
+    where p is the degree of the Lagrange function space. Solved on the
+    gdim-dimensional UnitMesh, with homogeneous Dirichlet boundary
+    conditions.
+
+    Compares values at each dof with the analytical solution.
+
+    """
+    if gdim == 1:
+        mesh = UnitIntervalMesh(MPI.comm_world, 10)
+    if gdim == 2:
+        mesh = UnitSquareMesh(MPI.comm_world, 15, 15, cell_type)
+    elif gdim == 3:
+        mesh = UnitCubeMesh(MPI.comm_world, 10, 10, 10, cell_type)
+    V = FunctionSpace(mesh, ("CG", degree))
+    u, v = TrialFunction(V), TestFunction(V)
+    x = SpatialCoordinate(mesh)
+
+    man_sol = ManufacturedSolutionPoisson(degree, gdim)
+    f = -eval(str(man_sol.laplace()))
+    a = inner(grad(u), grad(v)) * dx
+    L = inner(f, v) * dx
+
+    u_bc = Function(V)
+    with u_bc.vector.localForm() as u_local:
+        u_local.set(0.0)
+    bc = DirichletBC(V, u_bc, boundary)
+
+    uh = Function(V)
+    solve(a == L, uh, bc)
+    x = V.tabulate_dof_coordinates()
+    x_ = np.zeros((x.shape[0], 3))
+    # Fill with coordinates to work with eval
+    if gdim == 1:
+        x_[:, :1] = x
+    elif gdim == 2:
+        x_[:, :2] = x
+    else:
+        x_ = x
+
+    for i in range(V.mesh.num_cells()):
+        # Find points of dofs in i-th cell
+        dofs_i = V.dofmap.cell_dofs(i)
+        x_i = np.array([x_[dof, :] for dof in dofs_i])
+
+        # Function values for dofs at i-th cell
+        result_i = uh.eval(x_i, i * np.ones(len(x_i)))
+
+        # Compute exact solution
+        if mesh.geometric_dimension() == 1:
+            exact_i = man_sol.eval(x_i[:, 0], 0, 0)
+        elif mesh.geometric_dimension() == 2:
+            exact_i = man_sol.eval(x_i[:, 0], x_i[:, 1], 0)
+        elif mesh.geometric_dimension() == 3:
+            exact_i = man_sol.eval(x_i[:, 0], x_i[:, 1], x_i[:, 2])
+        # Reshape
+        result_i = result_i.reshape(exact_i.shape)
+        # Measure absolute error as values close to zero breaks the rtol.
+        assert max(exact_i - result_i) < 1e-3
```

Comment:
Do not use python std library `max()` on larger arrays. It is too generic and very slow. Prefer `numpy.max(a,b)`.